messaging/src/main/java/io/lantern/messaging/Messaging.kt

```diff
@@ -0,0 +1,821 @@
+                        .setSent(now)
+                        .setSenderId(myId.id)
+                        .setRecipientId(msg.contactId.id) // TODO: this will need to change for groups
+                tx.put(out.dbPath, out.build())
+                cryptoWorker.submit { cryptoWorker.processOutbound(out) }
+            }
+        }
+    }
+
+    /**
+     * Deletes the message locally only.
+     *
+     * @param msgPath the path identifying the message to delete.
+     */
+    fun deleteLocally(msgPath: String): Model.StoredMessage? {
+        return db.mutate { tx ->
+            doDeleteLocally(tx, msgPath)
+        }
+    }
+
+    private fun doDeleteLocally(tx: Transaction, msgPath: String): Model.StoredMessage? {
+        return db.get<Model.StoredMessage>(msgPath)?.let { msg ->
+            tx.delete(msgPath)
+
+            // Delete attachments on disk
+            // Note - we only delete attachments locally, not in the cloud, because clients
+            // aren't authorized to modify files. They will naturally be deleted once they hit
+            // the server-side retention limit.
+            msg.attachmentsMap.values.forEach { storedAttachment ->
+                if (!File(storedAttachment.encryptedFilePath).delete()) {
+                    logger.error("failed to delete attachment on disk, continuing")
+                }
+            }
+
+            // Delete index entries for messages under this Contact's conversation
+            tx.delete(msg.contactMessagePath)
+
+            // Update the Contact metadata based on the most recent remaining message
+            tx.listDetails<Model.StoredMessage>(
+                msg.contactMessagesQuery,
+                count = 1,
+                reverseSort = true
+            ).let { storedMessages ->
+                val mostRecentMsg = storedMessages.firstOrNull()
+                if (mostRecentMsg != null) {
+                    updateContactMetaData(tx, mostRecentMsg.value, force = true)
+                } else {
+                    clearContactMetaData(tx, msg.contactId)
+                }
+            }
+
+            return@let msg
+        }
+    }
+
+    /**
+     * Creates a StoredAttachment from the given File.
+     *
+     * @param file the File from which to create the attachment
+     * @param mimeType the mime type to use if it cannot be auto-detected
+     * @param metadata arbitrary metadata to associate with the attachment
+     * @param lazy if true, we won't encrypt the attachment yet (will be encrypted upon sending)
+     */
+    fun createAttachment(
+        file: File,
+        mimeType: String? = null,
+        metadata: Map<String, String>? = null,
+        lazy: Boolean = true,
+    ): Model.StoredAttachment {
+        val md = try {
+            Metadata.analyze(file, mimeType)
+        } catch (t: Throwable) {
+            logger.error("couldn't analyze metadata: ${t.message}")
+            null
```

Comment:
Do we present any sort of error to the user if creating/storing an attachment fails?

Reply:
if creating the attachment fails (like if there's a problem writing to disk), we'll throw an exception that can be handled by the UI. This particular branch just means that we couldn't extract metadata for the attachment, which doesn't prevent us from creating the attachment, it just means we don't have any metadata. That particular case is not indicated to the user, which I think is okay since there's nothing they could do about it anyway.